File: app/src/main/java/com/cogu/spylook/model/entity/Contacto.java
package com.cogu.spylook.model.entity;


import androidx.room.Entity;
import androidx.room.PrimaryKey;

import java.time.LocalDate;

import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@NoArgsConstructor
@Entity(tableName = "contactos")
public class Contacto {
    @PrimaryKey(autoGenerate = true)
    private int id;
    private String nombre;
    private String alias;
    private int edad;
    private int foto;
    private LocalDate fechaNacimiento;
    private String ciudad;
    private String estado;
    private String pais;

    public Contacto(int id, String nombre, String nickMasConocido, LocalDate fechaNacimiento, String ciudad, String estado, String pais) {
        this.id = id;
        this.nombre = nombre;
        this.alias = nickMasConocido;
        this.fechaNacimiento = fechaNacimiento;
        this.edad = LocalDate.now().isAfter(
                fechaNacimiento.withYear(LocalDate.now().getYear())
                        .withDayOfYear(LocalDate.now().getDayOfYear() - 1))
                ? LocalDate.now().getYear() - fechaNacimiento.getYear()
                : LocalDate.now().getYear() - (fechaNacimiento.getYear() - 1);
        this.ciudad = ciudad;
        this.estado = estado;
        this.pais = pais;
    }
    public Contacto(String nombre, String nickMasConocido, LocalDate fechaNacimiento, String ciudad, String estado, String pais) {
        this.nombre = nombre;
        this.alias = nickMasConocido;
        this.fechaNacimiento = fechaNacimiento;
        this.edad = LocalDate.now().isAfter(
                fechaNacimiento.withYear(LocalDate.now().getYear())
                        .withDayOfYear(LocalDate.now().getDayOfYear() - 1))
                ? LocalDate.now().getYear() - fechaNacimiento.getYear()
                : LocalDate.now().getYear() - (fechaNacimiento.getYear() - 1);
        this.ciudad = ciudad;
        this.estado = estado;
        this.pais = pais;
    }
}

File: app/src/main/java/com/cogu/spylook/model/entity/Contacto.kt
package com.cogu.spylook.model.entity

import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import java.time.LocalDate

@Entity(
    tableName = "contactos",
    indices = [androidx.room.Index(value = ["alias"], unique = true)]
)
class Contacto(
    idAnotable: Int = 0,
    nombre: String,
    @JvmField
    var alias: String,
    @JvmField
    var fechaNacimiento: LocalDate,
    @JvmField
    var ciudad: String,
    @JvmField
    var estado: String,
    @JvmField
    var pais: String,
    @JvmField
    var colorFoto: Int,
): Anotable(idAnotable, nombre) {
    @Ignore
    var edad: Int = calcularEdad(fechaNacimiento)

    @Ignore
    constructor(
        nombre: String,
        alias: String,
        fechaNacimiento: LocalDate,
        ciudad: String,
        estado: String,
        pais: String,
        colorFoto: Int
    ) : this(
        idAnotable = 0,
        nombre = nombre,
        alias = alias,
        fechaNacimiento = fechaNacimiento,
        ciudad = ciudad,
        estado = estado,
        pais = pais,
        colorFoto = colorFoto
    )

    companion object {
        private fun calcularEdad(fechaNacimiento: LocalDate?): Int {
            return if (fechaNacimiento != null) {
                if (LocalDate.now().isAfter(
                        fechaNacimiento.withYear(LocalDate.now().year)
                    )
                ) {
                    LocalDate.now().year - fechaNacimiento.year
                } else {
                    (LocalDate.now().year - (fechaNacimiento.year))- 1
                }
            } else 0
        }
    }
}